# DMA Fall 19

**Note** : This entire lab will be manually evaluated.

Name : 'Daniel Martin'

Collaborator : 'Eric Jian, Ashwath Mohan, Lavanya Vijayan'

# Lab 4: Neural Networks #


In [0]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [10]:
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train.head()

--2019-10-01 18:05:27--  http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 105581 (103K) [text/csv]
Saving to: ‘lab_4_training.csv.1’

lab_4_training.csv. 100%[===================>] 103.11K   489KB/s    in 0.2s    

2019-10-01 18:05:28 (489 KB/s) - ‘lab_4_training.csv.1’ saved [105581/105581]

--2019-10-01 18:05:29--  http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 26523 (26K) [text/csv]
Saving to: ‘lab_4_test.csv.1’

lab_4_test.csv.1    100%[===================>]  

Unnamed: 0 gender  age    year  ... exercisehours  musiccds  playgames  watchtv
0        1303   male   20  second  ...           5.0      50.0        1.0     15.0
1          36   male   20   third  ...           4.0      10.0        0.0      1.0
2         489   male   22  fourth  ...           2.0     150.0        1.0     10.0
3        1415   male   19  second  ...           5.0     100.0        0.0      7.0
4         616   male   22  fourth  ...           7.0      10.0        0.0      5.0

[5 rows x 15 columns]

***
### Question 1###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of 'gender'. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives)

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [11]:
# YOUR CODE HERE
df_train['gender'].value_counts().max() / df_train['gender'].value_counts().sum()

0.5377358490566038

###ANSWER: 0.5377358490566038

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [12]:
# YOUR CODE HERE
df_test['gender'].value_counts().max() / df_test['gender'].value_counts().sum()

0.5226130653266332

###ANSWER: 0.5226130653266332

***
### Question 2 ###
Get started with Neural Networks.

   
Choose a NN implementation (eg: scikit-learn) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (sigmoid/logistc) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (lbfgs). It is possible to specify logistic in Sklearn MLPclassifier (Neural net).  

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the Height feature of the dataset to predict the Gender. You will have to change Gender to a 0 and 1 class. After training, use your trained model to predict the class using the height feature from the training set. What was the accuracy of this prediction?

In [13]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

df_train['gender'].replace(0, 'Male', inplace = True)
df_train['gender'].replace(1, 'Female', inplace = True)
df_test['gender'].replace(0, 'Male', inplace = True)
df_test['gender'].replace(1, 'Female', inplace = True)

X_train = df_train[['height']]
y_train = df_train['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train,y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

0.5377358490566038


###ANSWER: 0.5377358490566038

**Question 2.b**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the Height feature values from the test set. What is the accuracy of this model on the test set?

In [14]:
# YOUR CODE HERE
from sklearn.metrics import accuracy_score

X_test = df_test[['height']]
y_test = df_test['gender']

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.5226130653266332


###ANSWER: 0.5226130653266332

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the height feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature

In [15]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
import numpy as np

df_train['log_height'] = df_train['height'].apply(np.log)
df_test['log_height'] = df_test['height'].apply(np.log)

# df_train['gender'].replace(0, 'Male', inplace = True)
# df_train['gender'].replace(1, 'Female', inplace = True)
# df_test['gender'].replace(0, 'Male', inplace = True)
# df_test['gender'].replace(1, 'Female', inplace = True)

X_train = df_train[['log_height']]
y_train = df_train['gender']

X_test = df_test[['log_height']]
y_test = df_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train,y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))


0.5616352201257861
0.5452261306532663


###ANSWER: 
Train Set: 0.5616352201257861

Test Set: 0.5452261306532663
           

***

### Question 3###
The rest of features in this dataset barring a few are categorical. No ML method accepts categorical features, so transform year, eyecolor, exercise into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [16]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

copy_df_train = df_train.copy()
copy_df_test = df_test.copy()

copy_df_train['year'].replace('first"', 'first', inplace = True)
copy_df_test['year'].replace('first"', 'first', inplace = True)
#copy_df_train.head()

df_one_hot_train = pd.get_dummies(copy_df_train, columns = ['year', 'eyecolor', 'exercise'])
df_one_hot_test = pd.get_dummies(copy_df_test, columns = ['year', 'eyecolor', 'exercise'])

X_train = df_one_hot_train.loc[:, "year_first":]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test.loc[:, "year_first":]
y_test = df_one_hot_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 50, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.5610062893081761
0.5879396984924623


###ANSWER: 
Test Set: 0.5879396984924623

***
### Question 4###
Using a NN, report the accuracy on  the test set of a model that trained only on the height and the eye color features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original height values (no pre-processing) and eye color as a one-hot?

In [17]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

df_one_hot_train = pd.get_dummies(df_train, columns = ['eyecolor'])
df_one_hot_test = pd.get_dummies(df_test, columns = ['eyecolor'])

X_train = df_one_hot_train.loc[:, ['height', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test.loc[:, ['height', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_test = df_one_hot_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.7679245283018868
0.7788944723618091


###ANSWER: 
Test Set: 0.7788944723618091

**Question 4.b**  
What is the accuracy on the test set using the log of height values (applied to both training and testing sets) and eye color as a one-hot?

In [18]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

df_one_hot_train = pd.get_dummies(df_train, columns = ['eyecolor'])
df_one_hot_test = pd.get_dummies(df_test, columns = ['eyecolor'])

X_train = df_one_hot_train.loc[:, ['log_height', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test.loc[:, ['log_height', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_test = df_one_hot_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.650314465408805
0.6507537688442211


###ANSWER: 
Train Set: 0.650314465408805

Test Set: 0.6507537688442211

**Question 4.c**  
What is the accuracy on the test set using the Z-score of height values and eye color as a one-hot? 

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [107]:
# YOUR CODE HERE
from scipy import stats

df_train['z_score_heights'] = stats.zscore(df_train['height'])
df_test['z_score_heights'] = stats.zscore(df_test['height'])

df_one_hot_train = pd.get_dummies(df_train, columns = ['eyecolor'])
df_one_hot_test = pd.get_dummies(df_test, columns = ['eyecolor'])


X_train = df_one_hot_train.loc[:, ['z_score_heights', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test.loc[:, ['z_score_heights', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_test = df_one_hot_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.8503144654088051
0.8592964824120602


###ANSWER: 
Train Set: 0.8503144654088051

Test Set: 0.8592964824120602

***
### Question 5 ###
Repeat question 4 for exercise hours + eye color

In [111]:
# YOUR CODE HERE
from scipy import stats

df_train['z_score_exercisehours'] = stats.zscore(df_train['exercisehours'])
df_test['z_score_exercisehours'] = stats.zscore(df_test['exercisehours'])

df_one_hot_train = pd.get_dummies(df_train, columns = ['eyecolor'])
df_one_hot_test = pd.get_dummies(df_test, columns = ['eyecolor'])


X_train = df_one_hot_train.loc[:, ['z_score_exercisehours', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test.loc[:, ['z_score_exercisehours', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other']]
y_test = df_one_hot_test['gender']

clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))

0.5937106918238994
0.5778894472361809


###ANSWER: 
Train Set: 0.5937106918238994

Test Set: 0.5778894472361809

***
### Question 6###
Combine the features from question 3, 4, and 5 (year, eyecolor, exercise, height, exercise hours). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [120]:
# YOUR CODE HERE
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

df_train['year'].replace('first"', 'first', inplace = True)
df_test['year'].replace('first"', 'first', inplace = True)

df_train['z_score_heights'] = stats.zscore(df_train['height'])
df_test['z_score_heights'] = stats.zscore(df_test['height'])
df_train['z_score_exercisehours'] = stats.zscore(df_train['exercisehours'])
df_test['z_score_exercisehours'] = stats.zscore(df_test['exercisehours'])

df_one_hot_train = pd.get_dummies(df_train, columns = ['year', 'eyecolor', 'exercise'])
df_one_hot_test = pd.get_dummies(df_test, columns = ['year', 'eyecolor', 'exercise'])
                                                
X_train = df_one_hot_train[['z_score_heights', 'z_score_exercisehours', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other', 'year_first', 'year_second', 'year_third', 'year_fourth', 'year_other', 'exercise_Yes', 'exercise_No']]
y_train = df_one_hot_train['gender']

X_test = df_one_hot_test[['z_score_heights', 'z_score_exercisehours', 'eyecolor_blue', 'eyecolor_brown', 'eyecolor_green', 'eyecolor_hazel', 'eyecolor_other', 'year_first', 'year_second', 'year_third', 'year_fourth', 'year_other', 'exercise_Yes', 'exercise_No']]
y_test = df_one_hot_test['gender']              
                                                
clf = MLPClassifier(hidden_layer_sizes = (10,2), max_iter = 100, alpha = 0.0001,
                   activation = 'logistic', solver = 'lbfgs', verbose = 10, 
                   random_state=42, tol=0.000000001)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
print(accuracy_score(y_train, y_train_pred))

y_test_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_test_pred))                                                

0.8660377358490566
0.8793969849246231


###ANSWER: 
Test Set: 0.8793969849246231

***
### Question 7- Bonus (10%)###
Can you improve your test set prediction accuracy by 5% or more?  

See how close to that milestone of improvement you can get by modifying the tuning parameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate aka mu). A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf. 

While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies to generally to any ML technique with tuning parameters.

Please also write a paragraph in a markdown cell below with an explanation of your approach and evaluation metrics.


In [0]:
# YOUR CODE HERE

###ANSWER: 